rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala

```diff
@@ -288,28 +291,36 @@ object Reduce {
           )
         )
 
-      @annotation.tailrec
-      def firstMatch(target: Par, cases: Seq[MatchCase])(implicit env: Env[Par]): M[Unit] =
-        cases match {
-          case Nil => Applicative[M].pure(())
-          case singleCase +: caseRem =>
-            val matchResult = SpatialMatcher
-              .spatialMatch(target, singleCase.pattern.get)
-              .runS(SpatialMatcher.emptyMap)
-            matchResult match {
-              case None => firstMatch(target, caseRem)
-              case Some(freeMap) => {
-                val newEnv: Env[Par] = addToEnv(env, freeMap, singleCase.freeCount)
-                eval(singleCase.source.get)(newEnv)
+      def firstMatch(target: Par, cases: Seq[MatchCase])(implicit env: Env[Par]): M[Unit] = {
```

Comment:
The problem was that we needed to substitute into the pattern now. That means that it's not tail-recursive now, so I wrote it so that monadicly tail-recursive. See line 301